src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs

```diff
@@ -70,6 +70,129 @@ public static NuGetFramework Parse(string folderName, IFrameworkNameProvider map
+        }
+
+        /// <summary>
+        /// Creates a NuGetFramework from individual components, using the given mappings
+        /// </summary>
+        public static NuGetFramework ParseComponents(string targetFrameworkIdentifier, string targetFrameworkVersion, string targetFrameworkProfile, string targetPlatformIdentifier, string targetPlatformVersion, IFrameworkNameProvider mappings)
+        {
+            if (targetFrameworkIdentifier == null)
+            {
+                throw new ArgumentNullException("targetFrameworkIdentifier");
+            }
+
+            if (mappings == null)
+            {
+                throw new ArgumentNullException("mappings");
+            }
+
+            NuGetFramework result = null;
+
+            if (!TryParseSpecialFramework(targetFrameworkIdentifier, out result))
+            {
+                string framework = null;
+                if (!mappings.TryGetIdentifier(targetFrameworkIdentifier, out framework))
+                {
+                    framework = targetFrameworkIdentifier;
+                }
+
+                var version = new Version(0, 0);
+                var platformVersion = new Version(0, 0);
+                string profile = null;
+
+                if (!string.IsNullOrEmpty(targetFrameworkVersion))
+                {
+                    if (targetFrameworkVersion.IndexOf('.') < 0)
+                    {
+                        targetFrameworkVersion += ".0";
+                    }
+
+                    if (!Version.TryParse(targetFrameworkVersion, out version))
+                    {
+                        throw new ArgumentException(string.Format(
+                            CultureInfo.CurrentCulture,
+                            Strings.InvalidFrameworkVersion,
+                            targetFrameworkVersion));
+                    }
+                }
+
+                if (version.Major >= 5
+                    && StringComparer.OrdinalIgnoreCase.Equals(FrameworkConstants.FrameworkIdentifiers.Net, framework))
+                {
+                    framework = FrameworkConstants.FrameworkIdentifiers.NetCoreApp;
+                }
+
+                if (StringComparer.OrdinalIgnoreCase.Equals(FrameworkConstants.FrameworkIdentifiers.Portable, framework)
+                    && !string.IsNullOrEmpty(profile)
+                    && profile.Contains("-"))
+                {
+                    // Frameworks within the portable profile are not allowed
+                    // to have profiles themselves #1869
+                    throw new ArgumentException(string.Format(
+                        CultureInfo.CurrentCulture,
+                        Strings.InvalidPortableFrameworksDueToHyphen,
+                        profile));
+                }
+
+                if (!mappings.TryGetProfile(framework, targetFrameworkProfile ?? string.Empty, out profile))
+                {
+                    profile = targetFrameworkProfile ?? string.Empty;
```

Comment:
Shouldn't coalesce on empty string.  `null` (not a zero-length string) is how to represent the absence of a value.  The consumer (`NuGetFramework`'s constructor) handles `null` profile just fine.

(I see that `NuGetFramework` is internally inconsistent with respect to `null` handling of platform and profile, but that's life.)